content/blog/2019-10-26-lazy-code-motion/index.md

```diff
@@ -0,0 +1,206 @@
+the Ocamlgraph framework made it hard to refer to results of other dataflow
+passes in later ones. I worked around this by tagging every basic block in the
+CFG with a map of "attributes" and writing Ocaml functors like `MakeDataflow`
+that require their input module to include an `attr_name` field. When you run
+a dataflow analysis created with `MakeDataflow`, it saves the results of the
+analysis at each basic block into the attributes map under the `attr_name` key.
+
+The analyses were easy to write down once there was structure in place for
+defining them. For example, here is the definition of an available expressions
+analysis from the paper:
+
+```
+let avail_out (_, block_attrs) avail_in =
+  Bitv.bw_or
+    (Bitv.bw_and
+       (Cfg.Attrs.get block_attrs "transparent")
+       avail_in)
+    (Cfg.Attrs.get block_attrs "computes")
+
+module AvailabilityIn =
+  MakeDataflow
+    (struct
+      let attr_name = "availability_in"
+      let direction = Graph.Fixpoint.Forward
+      let init _ =
+        EXPRS.build ~f:(fun _ -> false)
+      let analyze (src, _, _) src_avail_in =
+        avail_out src src_avail_in
+    end)
+```
+
+I implemented a CFG printer using the Ocamlgraph Graphviz backend so that
+I could debug the optimization's changes to CFGs directly, rather than staring
+at pretty printed output. Here's the result of running the optimization on
+`tests/lcm/simple`, a nested loop benchmark. There's a way to include the
+instructions from each block inside the nodes, but for this example that makes
+it hard to follow. There is a loop invariant computation in `inner_body` that
+lazy code motion hoists to `lcm_inserted_block0`.
+
+<img src="simple-before.png" />
+<img src="simple-after.png" />
+
+# Limitations
+I implemented the full lazy code motion optimization, but there were some
+issues. The optimization expects lexically equal expressions to always be stored
+into the same pseudoregister. There may be adaptations of the dataflow analyses
+in later work to weaken this assumption, but I introduced a conservative code
+rewriting pass that made sure expressions went into a uniquely named temporary.
+This increases register pressure and introduces extraneous move instructions to
+retrieve computed values from temporaries. A more clever rewriting pass might be
+able to ameliorate these costs.
+
+On the other end of the optimization, the computation placement algorithm is
+inefficient. Lazy code motion places computations on edges of the CFG, which
+means stitching new basic blocks into edges. While they are neccessary in
+general, on many edges an inserted basic block could be safely merged with its
+predecessor or successor block. This could have performance benefits because it
+would reduce the number of jumps. Similarly the pretty-printer for CFGs
+does not omit jumps where fall-through would work--this seems like a minor
+detail but could have performance or code size implications.
```

Comment:
In an ideal world, there would be an off-the-shelf "CFG simplification" pass that you could run after your pass to clean up these easily-coalescable cases.